docsite/docs/Docs/fal-cli/environments.md

```diff
@@ -0,0 +1,298 @@
+    meta:
+      fal:
+        environment: interact-cloud
+        pre-hook:
+          - path: change_s3_permissions.py
+            with:
+              can_dbt_read: yes
+
+        post-hook:
+          - path: change_s3_permissions.py
+            with:
+              can_dbt_read: no
+          - path: send_slack_alert.py
+            environment: alerts
+```
+
+If we add `send_slack_alert.py` to the example above, all the `s3` related
+scripts will still run in the `interact-cloud` environment but the
+`send_slack_alert.py` will now run under the `alerts` environment.
+
+#### Overriding the model-level environment with the local environment
+
+If you want your hooks to run in the same environment as your `fal` process
+(without any sort of influence from the outside scopes, e.g. model-level
+environments), you can use a reserved environment called `local`. It makes Fal
+run them as if they are regular hooks without any environments, in the same
+Python runtime that runs Fal itself.
+
+
+```yml
+version: 2
+
+models:
+  - name: load_data_from_s3
+    meta:
+      fal:
+        environment: interact-cloud
+        pre-hook:
+          - path: print_enter_banner.py
+            environment: local
+
+        post-hook:
+          - path: print_exit_banner.py
+            environment: local
+```
+
+The `load_data_from_s3.py` (Fal model) will still continue to be ran inside
+`interact-cloud` environment, but all the banner printing hooks now specify
+`local` as their environment (which is a reserved name, so you can't re-define
+it). Fal will run them as if it were running them outside of this model-level
+environment's scope.
+
+
+## Environment caches
+
+Since creating an environment is the longest part of the job, Fal will avoid
+doing so as long as there weren't any changes in the environment's definition.
+
+```yml
+environments:
+  - name: alerts
+    type: venv
+    requirements:
+      - slack-sdk==1.1.0
+
+  - name: training
+    type: conda
+    packages:
+      - scikit-learn=2.0.0
+```
+
+When the `alerts` or `training` environment is used for the first time, Fal will
+create it from scratch and save it under your user cache dir (this depends on
+your system, but you can see the default
+[value in this page](https://github.com/platformdirs/platformdirs#example-output)).
+For all the subsequent runs (of either the same script, or different ones that
+use the same environment) Fal will try to use the already created environment
+unless the environment definition itself has been changed. If the
+`fal_project.yml` above evolves into something like the following (e.g. a new
+dependency has been added under the `training` environment):
+
+```yml
+environments:
+  - name: alerts
+    type: venv
+    requirements:
+      - slack-sdk==1.1.0
+
+  - name: training
+    type: conda
+    packages:
+      - scikit-learn=2.0.0
+      - xgboost=1.0.0
+```
+
+Fal will still use the same `alerts` environment no matter what, since it can
```

Comment:
Do we want to say anything about cleaning up environments? Maybe some example conda or venv commands?